src/de/deserializer.zig

```diff
@@ -598,33 +546,37 @@ fn Union(comptime D: type) type {
 
         const Error = De.Error;
 
-        fn variantSeed(self: *Self, _: ?std.mem.Allocator, seed: anytype) Error!@TypeOf(seed).Value {
+        fn variantSeed(self: *Self, allocator: ?std.mem.Allocator, seed: anytype) Error!@TypeOf(seed).Value {
             comptime concepts.Concept("StringVariant", "expected variant type to be a string")(.{
                 concepts.traits.isString(@TypeOf(seed).Value),
             });
 
-            const token = (try self.d.tokens.next()) orelse return error.MissingVariant;
+            const token = try self.d.tokens.nextAlloc(allocator orelse return error.MissingAllocator, .alloc_if_needed);
+            defer self.d.freeToken(token);
 
-            if (token == .String) {
-                return token.String.slice(self.d.tokens.slice, self.d.tokens.i - 1);
-            }
+            switch (token) {
+                .end_of_document => return error.MissingVariant,
+                .string => |s| return s,
+                .allocated_string => |s| {
+                    if (s.len > self.variant_buf.len)
+                        return error.OutOfMemory;
+                    @memcpy(&self.variant_buf, s);
+                    return self.variant_buf[0..s.len];
+                },
 
-            return error.InvalidType;
+                else => return error.InvalidType,
+            }
         }
 
         fn payloadSeed(self: *Self, allocator: ?std.mem.Allocator, seed: anytype) Error!@TypeOf(seed).Value {
-            if (@TypeOf(seed).Value != void) {
+            if (comptime @TypeOf(seed).Value != void) {
```

Comment:
`comptime` is redundant here.

```suggestion
            if (@TypeOf(seed).Value != void) {
```